File: frontend/src/services/exam.ts
import api from './api';

export interface ExamStartRequest {
    subject_id: string;
    topic_ids: string[];
    num_questions: number;
    time_limit_minutes?: number;
}

export interface ExamQuestion {
    question_id: string;
    question: string;
    options: string[];
    topic_id: string;
    topic_name: string;
}

export interface ExamStartResponse {
    exam_id: string;
    subject_name: string;
    topics: { topic_id: string; topic_name: string }[];
    questions: ExamQuestion[];
    time_limit_seconds: number | null;
    total_questions: number;
}

export interface ExamSubmissionItem {
    question_id: string;
    question: string;
    options: string[];
    answer: string;
    correct_answer: string;
    topic_id: string;
}

export interface ExamSubmitRequest {
    exam_id: string;
    subject_id: string;
    subject_name: string;
    topic_ids: string[];
    answers: ExamSubmissionItem[];
    duration_seconds?: number;
}

export interface TopicBreakdown {
    topic_id: string;
    topic_name: string;
    correct: number;
    total: number;
    percentage: number;
}

export interface ExamFeedback {
    overall_interpretation: string;
    topic_analysis: string[];
    strengths: string[];
    areas_to_focus: string[];
    study_recommendations: string[];
    encouraging_message: string;
}

export interface ExamResult {
    id: string;
    score: number;
    total_questions: number;
    correct_questions: number;
    duration_seconds?: number;
    completed_at: string;
    subject_name: string;
    topic_breakdown: TopicBreakdown[];
    feedback?: ExamFeedback;
}

export interface ExamHistoryItem {
    id: string;
    score: number;
    total_questions: number;
    correct_questions: number;
    subject_name: string;
    topics_count: number;
    completed_at: string;
}

export const examService = {
    /**
     * Start a new exam
     */
    async startExam(data: ExamStartRequest): Promise<ExamStartResponse> {
        const response = await api.post<ExamStartResponse>('/exams/start', data);
        return response.data;
    },

    /**
     * Submit an exam
     */
    async submitExam(data: ExamSubmitRequest): Promise<ExamResult> {
        const response = await api.post<ExamResult>('/exams/submit', data);
        return response.data;
    },

    /**
     * Get exam history
     */
    async getHistory(): Promise<ExamHistoryItem[]> {
        const response = await api.get<ExamHistoryItem[]>('/exams/history');
        return response.data;
    }
};

export default examService;

File: backend/app/schemas/exam.py
"""
AI Tutor Platform - Exam Schemas
Pydantic schemas for subject-level exam requests and responses
"""
import uuid
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field


class ExamTopicSelection(BaseModel):
    """A topic selected for the exam."""
    topic_id: uuid.UUID
    topic_name: str


class ExamStartRequest(BaseModel):
    """Request to start an exam."""
    subject_id: uuid.UUID
    topic_ids: list[uuid.UUID] = Field(..., min_length=1, description="List of topic IDs to include")
    num_questions: int = Field(default=10, ge=5, le=50, description="Number of questions (5-50)")
    time_limit_minutes: int | None = Field(default=None, ge=5, le=120, description="Optional time limit in minutes")


class ExamQuestion(BaseModel):
    """A single question in an exam."""
    question_id: str
    question: str
    options: list[str]
    topic_id: str  # Track which topic this question belongs to
    topic_name: str


class ExamStartResponse(BaseModel):
    """Response when starting an exam."""
    exam_id: str
    subject_name: str
    topics: list[ExamTopicSelection]
    questions: list[ExamQuestion]
    time_limit_seconds: int | None = None
    total_questions: int


class ExamSubmissionItem(BaseModel):
    """Single answer submission with question context."""
    question_id: str
    question: str
    options: list[str]
    answer: str
    correct_answer: str
    topic_id: str


class ExamSubmitRequest(BaseModel):
    """Request to submit an exam."""
    exam_id: str
    subject_id: uuid.UUID
    subject_name: str
    topic_ids: list[uuid.UUID]
    answers: list[ExamSubmissionItem]
    duration_seconds: int | None = None  # Time taken by student


class TopicBreakdown(BaseModel):
    """Score breakdown for a single topic."""
    topic_id: str
    topic_name: str
    correct: int
    total: int
    percentage: float


class ExamFeedbackResponse(BaseModel):
    """AI-generated detailed feedback for an exam."""
    overall_interpretation: str
    topic_analysis: list[str]  # Per-topic insights
    strengths: list[str]
    areas_to_focus: list[str]
    study_recommendations: list[str]
    encouraging_message: str


class ExamResultResponse(BaseModel):
    """Response after submitting an exam."""
    model_config = ConfigDict(from_attributes=True)
    
    id: uuid.UUID
    score: float
    total_questions: int
    correct_questions: int
    duration_seconds: int | None = None
    completed_at: datetime
    subject_name: str
    topic_breakdown: list[TopicBreakdown]
    feedback: ExamFeedbackResponse | None = None


class ExamHistoryItem(BaseModel):
    """Summary of a past exam for history listing."""
    model_config = ConfigDict(from_attributes=True)
    
    id: uuid.UUID
    score: float
    total_questions: int
    correct_questions: int
    subject_name: str
    topics_count: int
    completed_at: datetime
